components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java

```diff
@@ -1505,6 +1510,93 @@ public static SpOAuth2ExpiryTimeConfiguration getSpTokenExpiryTimeConfig(String
         return spTokenTimeObject;
     }
 
+    /**
+     * Retrieve audience configured for the particular service provider.
+     *
+     * @param clientId
+     * @param tenantDomain
+     * @return
+     */
+    public static List<String> getOIDCAudience(String clientId, String tenantDomain) {
+
+        List<String> oidcAudiences = getDefinedCustomOIDCAudiences(clientId, tenantDomain);
+        // Need to add client_id as an audience value according to the spec.
+        if (!oidcAudiences.contains(clientId)) {
+            oidcAudiences.add(0, clientId);
+        } else {
+            Collections.swap(oidcAudiences, oidcAudiences.indexOf(clientId), 0);
+        }
+        return oidcAudiences;
+    }
+
+    private static List<String> getDefinedCustomOIDCAudiences(String clientId, String tenantDomain) {
+
+        List<String> audiences = new ArrayList<>();
+
+        // Priority should be given to service provider specific audiences over globally configured ones.
+        if (OAuth2ServiceComponentHolder.isAudienceEnabled()) {
+            OIDCAudienceCacheKey cacheKey = new OIDCAudienceCacheKey(clientId + "@" + tenantDomain);
+            CacheEntry result = OIDCAudienceCache.getInstance().getValueFromCache(cacheKey);
+
+            // Check cache hit, do the type check.
+            if (result != null && result instanceof OIDCAudienceCacheEntry) {
+                audiences = ((OIDCAudienceCacheEntry) result).getAudiences();
+                return audiences;
+            } else {
+                OAuthAppDAO oAuthAppDAO = new OAuthAppDAO();
+                try {
+                    audiences = oAuthAppDAO.getOIDCAudiences(tenantDomain, clientId);
```

Comment:
Can't we get the appDO and get the audience from that?
The reason why I propose this that, this will be an additional DB call that we can avoid. 
If we use the method in OAuth2Util to fetch the OAuth app, then we can take advantage of caching as well.

Reply:
All the places that use the `getOIDCAudience` already have the `OAuthAppDO` already retrieved. So passing that object to the `getOIDCAudience` and then using it from `getDefinedCustomOIDCAudiences` would be a better way without retrieving it again.